### DISCUSSION TITLE
How to concurrently receive and send data via WebSocketConn? How to process message frames?

### DISCUSSION THREAD

**Discussion by quixoticaxis** [trillium-rs/trillium#201] [Q&A] [answered]
Hello.

I'm probably missing something, but `WebSocketConn` does not implement `Clone` while `send()` and `next()` use `&mut self`, so sharing the reference is not an option. How to receive and send data concurrently?

I'm wondering if there is a way to receive and send data frame by frame? It seems that `send()` and `next()` operate with `Message` abstraction, which defines already assembled whole message. Also it's not clear how to send `Ping` in the middle of a message, or, more realistic, receive `Pong` in the middle of the message.

**Comment by jbr** [comment, marked as answer]
You can use [`WebSocketConn::take_inbound_stream`](https://docs.rs/trillium-websockets/latest/trillium_websockets/struct.WebSocketConn.html#method.take_inbound_stream) to concurrently send and receive messages.

For non-message-based websockets, it sounds like you might need a different websocket adapter than the one provided by trillium-websockets, or alternatively, a way to drop down to async-tungstenite. Does [`async_tungstenite::WebSocketStream`](https://docs.rs/async-tungstenite/0.17.2/async_tungstenite/struct.WebSocketStream.html) provide the affordances you need? It should be possible to provide a `impl TryFrom<WebSocketConn> for WebSocketStream`

**Reply by jbr** [reply]
If the async_tungstenite::WebSocketStream doesn't support your use case, you'd need to find or write a different async websocket protocol implementation. I'd happily help write another websocket handler for trillium that wraps an existing library, but currently I'm not looking to maintain a websocket protocol implementation just for trillium.

**Reply by jbr** [reply]
It might be worth adding that the inner library (tungstenite) [handles ping/pong interactions automatically](https://github.com/snapview/tungstenite-rs/blob/1b999136efb8d8349d2b461e28c3f5c0c106ea4c/src/protocol/mod.rs#L172-L175)

**Reply by quixoticaxis** [reply]
Thank you.
I doubt I would be able to handle the scenario, because only [tokio-tungstenite](https://docs.rs/async-tungstenite/0.17.2/async_tungstenite/struct.WebSocketStream.html) supports the split into separate stream and sink components, the runtime agnostic library does not.

**Reply by jbr** [reply]
Trillium's WebSocketConn splits the stream and sink internally. When you take_inbound_stream, that's taking the stream side and leaving the sink side on the Conn.

**Reply by jbr** [reply]
I consider it a design flaw of async-tungstenite's that they chose to offer a single Sink+Stream type that basically requires splitting it to do anything useful. Both trillium's WebSocketConn and tide's [WebSocketConnection](https://docs.rs/tide-websockets/latest/tide_websockets/struct.WebSocketConnection.html) (which I also authored) have to work around that awkward api

**Reply by quixoticaxis** [reply]
Oh, I see. Sorry, didn't understand it from the docs. I'll try it later.
